controllers/bmcjob_controller.go

```diff
@@ -18,38 +18,207 @@ package controllers
 
 import (
 	"context"
+	"fmt"
+	"strconv"
 
-	"k8s.io/apimachinery/pkg/runtime"
+	"github.com/go-logr/logr"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/log"
 
 	bmcv1alpha1 "github.com/tinkerbell/rufio/api/v1alpha1"
 )
 
 // BMCJobReconciler reconciles a BMCJob object
 type BMCJobReconciler struct {
-	client.Client
-	Scheme *runtime.Scheme
+	client           client.Client
+	bmcClientFactory BMCClientFactoryFunc
+	bmcTaskHandler   BMCTaskHandler
+	logger           logr.Logger
+}
+
+// NewBMCJobReconciler returns a new BMCJobReconciler
+func NewBMCJobReconciler(client client.Client, bmcClientFactory BMCClientFactoryFunc, bmcTaskHandler BMCTaskHandler, logger logr.Logger) *BMCJobReconciler {
+	return &BMCJobReconciler{
+		client:           client,
+		bmcClientFactory: bmcClientFactory,
+		bmcTaskHandler:   bmcTaskHandler,
+		logger:           logger,
+	}
 }
 
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=bmcjobs,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=bmcjobs/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=bmcjobs/finalizers,verbs=update
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the BMCJob object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+// Reconcile runs a BMCJob.
+// Creates the individual BMCTasks on the cluster.
+// Runs each Task in a BMCTask and updates the BMCJob status.
 func (r *BMCJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := r.logger.WithValues("BMCJob", req.NamespacedName)
+	logger.Info("Reconciling BMCJob")
+
+	// Fetch the bmcJob object
+	bmcJob := &bmcv1alpha1.BMCJob{}
+	err := r.client.Get(ctx, req.NamespacedName, bmcJob)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "Failed to get BMCJob")
+		return ctrl.Result{}, err
+	}
+
+	// Deletion is a noop.
+	if !bmcJob.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	// Job has StartTime is noop.
+	if !bmcJob.Status.StartTime.IsZero() {
```

Comment:
Is the design of this controller to _hold_ the reconciliation goroutine while tasks execute? If so, its probably fine for today but we might consider not doing that in the future and instead lean into watching tasks from the job controller and using owner references to retrieve the right job, figure out the next step and execute. There are complications with that approach too around client caching so will need some discussion.